File: packages/core/src/storage.ts
import Database from 'better-sqlite3';
import path from 'path';
import fs from 'fs';
import os from 'os';
import type { MemoryEntry, SearchResult, ContextKeeperConfig, DecisionHistory } from './types';

const SCHEMA = `
  CREATE TABLE IF NOT EXISTS memory (
    id            INTEGER PRIMARY KEY AUTOINCREMENT,
    type          TEXT NOT NULL DEFAULT 'note',
    title         TEXT NOT NULL,
    content       TEXT NOT NULL,
    tags          TEXT NOT NULL DEFAULT '[]',
    source        TEXT,
    status        TEXT NOT NULL DEFAULT 'active',
    supersedes_id INTEGER REFERENCES memory(id),
    change_reason TEXT,
    created_at    TEXT NOT NULL DEFAULT (datetime('now')),
    updated_at    TEXT NOT NULL DEFAULT (datetime('now'))
  );

  CREATE VIRTUAL TABLE IF NOT EXISTS memory_fts USING fts5(
    title, content, tags,
    content='memory',
    content_rowid='id'
  );

  CREATE TRIGGER IF NOT EXISTS memory_ai AFTER INSERT ON memory BEGIN
    INSERT INTO memory_fts(rowid, title, content, tags)
    VALUES (new.id, new.title, new.content, new.tags);
  END;

  CREATE TRIGGER IF NOT EXISTS memory_au AFTER UPDATE ON memory BEGIN
    INSERT INTO memory_fts(memory_fts, rowid, title, content, tags)
    VALUES ('delete', old.id, old.title, old.content, old.tags);
    INSERT INTO memory_fts(rowid, title, content, tags)
    VALUES (new.id, new.title, new.content, new.tags);
  END;

  CREATE TRIGGER IF NOT EXISTS memory_ad AFTER DELETE ON memory BEGIN
    INSERT INTO memory_fts(memory_fts, rowid, title, content, tags)
    VALUES ('delete', old.id, old.title, old.content, old.tags);
  END;
`;

/** Types mirrored to workspace DB — cross-project architectural knowledge */
const WORKSPACE_MIRROR_TYPES: MemoryEntry['type'][] = ['decision', 'pattern', 'workspace', 'rejected'];

export class ContextKeeperStorage {
  private db: Database.Database;
  private globalDb?: Database.Database;
  private workspaceDb?: Database.Database;
  private config: ContextKeeperConfig;

  constructor(config: ContextKeeperConfig) {
    this.config = config;
    const dir = path.dirname(config.dbPath);
    if (!fs.existsSync(dir)) {
      fs.mkdirSync(dir, { recursive: true });
    }
    this.db = new Database(config.dbPath);
    this.db.pragma('journal_mode = WAL');
    this.db.exec(SCHEMA);
    this.migrate();

    if (config.globalDbPath) {
      const globalDir = path.dirname(config.globalDbPath);
      if (!fs.existsSync(globalDir)) fs.mkdirSync(globalDir, { recursive: true });
      this.globalDb = new Database(config.globalDbPath);
      this.globalDb.pragma('journal_mode = WAL');
      this.globalDb.exec(SCHEMA);
    }

    if (config.workspaceDbPath) {
      const wsDir = path.dirname(config.workspaceDbPath);
      if (!fs.existsSync(wsDir)) fs.mkdirSync(wsDir, { recursive: true });
      this.workspaceDb = new Database(config.workspaceDbPath);
      this.workspaceDb.pragma('journal_mode = WAL');
      this.workspaceDb.exec(SCHEMA);
    }
  }

  /** Add new columns if upgrading from older DB schema */
  private migrate(): void {
    const cols = (this.db.prepare(`PRAGMA table_info(memory)`).all() as any[]).map(r => r.name);
    if (!cols.includes('status')) {
      this.db.exec(`ALTER TABLE memory ADD COLUMN status TEXT NOT NULL DEFAULT 'active'`);
    }
    if (!cols.includes('supersedes_id')) {
      this.db.exec(`ALTER TABLE memory ADD COLUMN supersedes_id INTEGER`);
    }
    if (!cols.includes('change_reason')) {
      this.db.exec(`ALTER TABLE memory ADD COLUMN change_reason TEXT`);
    }
  }

  add(entry: Omit<MemoryEntry, 'id' | 'created_at' | 'updated_at' | 'status'>): MemoryEntry {
    const stmt = this.db.prepare(`
      INSERT INTO memory (type, title, content, tags, source, supersedes_id, change_reason)
      VALUES (@type, @title, @content, @tags, @source, @supersedes_id, @change_reason)
    `);
    const params = {
      type: entry.type,
      title: entry.title,
      content: entry.content,
      tags: JSON.stringify(entry.tags),
      source: entry.source ?? null,
      supersedes_id: entry.supersedes_id ?? null,
      change_reason: entry.change_reason ?? null,
    };
    const result = stmt.run(params);

    // Mirror to global DB — tag with project root so cross-project search knows the origin
    if (this.globalDb) {
      try {
        this.globalDb.prepare(`
          INSERT INTO memory (type, title, content, tags, source, supersedes_id, change_reason)
          VALUES (@type, @title, @content, @tags, @source, @supersedes_id, @change_reason)
        `).run({ ...params, source: this.config.projectRoot });
      } catch { /* global DB write failure is non-fatal */ }
    }

    // Mirror workspace-relevant types to workspace DB
    if (this.workspaceDb && WORKSPACE_MIRROR_TYPES.includes(entry.type as MemoryEntry['type'])) {
      try {
        this.workspaceDb.prepare(`
          INSERT INTO memory (type, title, content, tags, source, supersedes_id, change_reason)
          VALUES (@type, @title, @content, @tags, @source, @supersedes_id, @change_reason)
        `).run({ ...params, source: this.config.projectRoot });
      } catch { /* workspace DB write failure is non-fatal */ }
    }

    return this.getById(result.lastInsertRowid as number)!;
  }

  /**
   * Update a decision: marks the old entry as 'superseded' and creates a new active entry.
   * Returns the new entry. The old entry is preserved in history.
   */
  updateDecision(
    supersedes_id: number,
    newEntry: Pick<MemoryEntry, 'title' | 'content' | 'tags'>,
    change_reason: string,
  ): MemoryEntry {
    const old = this.getById(supersedes_id);
    if (!old) throw new Error(`Entry #${supersedes_id} not found`);

    return this.db.transaction(() => {
      // Mark old entry as superseded
      this.db.prepare(`UPDATE memory SET status = 'superseded', updated_at = datetime('now') WHERE id = ?`).run(supersedes_id);
      // Create new entry that supersedes the old one
      return this.add({
        type: old.type,
        title: newEntry.title,
        content: newEntry.content,
        tags: newEntry.tags,
        supersedes_id,
        change_reason,
      });
    })();
  }

  getById(id: number): MemoryEntry | undefined {
    const row = this.db.prepare('SELECT * FROM memory WHERE id = ?').get(id) as any;
    return row ? this.deserialize(row) : undefined;
  }

  /**
   * Search only the global DB — returns entries from OTHER projects (excludes current project).
   */
  searchGlobal(query: string, limit = 10): SearchResult[] {
    if (!this.globalDb) return [];
    const sanitized = query.replace(/['"*]/g, ' ').trim();
    if (!sanitized) return [];

    const rows = this.globalDb.prepare(`
      SELECT m.*, memory_fts.rank AS score
      FROM memory_fts
      JOIN memory m ON memory_fts.rowid = m.id
      WHERE memory_fts MATCH ?
      AND m.status = 'active'
      AND (m.source IS NULL OR m.source != ?)
      ORDER BY rank
      LIMIT ?
    `).all(`${sanitized}*`, this.config.projectRoot, limit) as any[];

    return rows.map(row => ({ ...this.deserialize(row), score: Math.abs(row.score) }));
  }

  /**
   * Search the workspace DB — cross-project architectural knowledge.
   */
  searchWorkspace(query: string, limit = 10): SearchResult[] {
    if (!this.workspaceDb) return [];
    const sanitized = query.replace(/['"*]/g, ' ').trim();
    if (!sanitized) return [];

    const rows = this.workspaceDb.prepare(`
      SELECT m.*, memory_fts.rank AS score
      FROM memory_fts
      JOIN memory m ON memory_fts.rowid = m.id
      WHERE memory_fts MATCH ?
      AND m.status = 'active'
      ORDER BY rank
      LIMIT ?
    `).all(`${sanitized}*`, limit) as any[];

    return rows.map(row => ({ ...this.deserialize(row), score: Math.abs(row.score) }));
  }

  /**
   * Generate a compact markdown snapshot of workspace memory (cross-project entries).
   * Shows which project each entry came from.
   */
  generateWorkspaceContextMd(): string | null {
    if (!this.workspaceDb) return null;
    const rows = this.workspaceDb.prepare(
      `SELECT * FROM memory WHERE status = 'active' ORDER BY type, updated_at DESC LIMIT 200`
    ).all() as any[];
    if (rows.length === 0) return null;

    const entries = rows.map(r => this.deserialize(r));
    const grouped = entries.reduce((acc, e) => {
      if (!acc[e.type]) acc[e.type] = [];
      acc[e.type].push(e);
      return acc;
    }, {} as Record<string, MemoryEntry[]>);

    const order: MemoryEntry['type'][] = ['workspace', 'decision', 'pattern', 'rejected'];
    const sections = order
      .filter(t => grouped[t]?.length)
      .map(t => {
        const header = `### Workspace ${t.charAt(0).toUpperCase() + t.slice(1)}s`;
        const items = grouped[t].map(e => {
          const project = e.source ? path.basename(e.source) : 'shared';
          return `- **${e.title}** [id:${e.id}] [${project}]: ${e.content.split('\n')[0]}`;
        }).join('\n');
        return `${header}\n${items}`;
      });

    return sections.length ? sections.join('\n\n') : null;
  }

  /**
   * Search local, workspace, and global DBs, merging results.
   * Workspace first (most relevant cross-project), then local, then other projects.
   */
  searchAll(query: string, limit = 10): SearchResult[] {
    const local = this.search(query, limit);
    const workspace = this.searchWorkspace(query, Math.ceil(limit / 2));
    const global = this.searchGlobal(query, Math.ceil(limit / 2));
    return [...workspace, ...local, ...global]
      .sort((a, b) => b.score - a.score)
      .slice(0, limit);
  }

  /**
   * Search active memory entries using full-text search.
   */
  search(query: string, limit = 10, includeSuperseded = false): SearchResult[] {
    const sanitized = query.replace(/['"*]/g, ' ').trim();
    const statusFilter = includeSuperseded ? '' : `AND m.status = 'active'`;

    if (!sanitized) {
      return this.list(limit).map(e => ({ ...e, score: 1 }));
    }

    const rows = this.db.prepare(`
      SELECT m.*, memory_fts.rank AS score
      FROM memory_fts
      JOIN memory m ON memory_fts.rowid = m.id
      WHERE memory_fts MATCH ?
      ${statusFilter}
      ORDER BY rank
      LIMIT ?
    `).all(`${sanitized}*`, limit) as any[];

    return rows.map(row => ({
      ...this.deserialize(row),
      score: Math.abs(row.score),
    }));
  }

  /**
   * Get the full decision history for an entry (current + all superseded ancestors).
   */
  history(id: number): DecisionHistory | undefined {
    const current = this.getById(id);
    if (!current) return undefined;

    const historyEntries: MemoryEntry[] = [];
    let cursor: MemoryEntry | undefined = current;

    // Walk the supersedes chain backward
    while (cursor?.supersedes_id) {
      const prev = this.getById(cursor.supersedes_id);
      if (!prev) break;
      historyEntries.push(prev);
      cursor = prev;
    }

    return { current, history: historyEntries };
  }

  /**
   * List active entries, optionally filtered by type.
   */
  list(limit = 50, type?: MemoryEntry['type'], includeSuperseded = false): MemoryEntry[] {
    const statusFilter = includeSuperseded ? '' : `AND status = 'active'`;
    let rows;
    if (type) {
      rows = this.db.prepare(`SELECT * FROM memory WHERE type = ? ${statusFilter} ORDER BY updated_at DESC LIMIT ?`).all(type, limit);
    } else {
      rows = this.db.prepare(`SELECT * FROM memory WHERE 1=1 ${statusFilter} ORDER BY updated_at DESC LIMIT ?`).all(limit);
    }
    return (rows as any[]).map(row => this.deserialize(row));
  }

  delete(id: number): boolean {
    const result = this.db.prepare('DELETE FROM memory WHERE id = ?').run(id);
    return result.changes > 0;
  }

  /**
   * Save an auto-generated session summary (called by Stop hook).
   */
  addSessionSummary(content: string, tags: string[] = []): MemoryEntry {
    const date = new Date().toISOString().split('T')[0];
    const time = new Date().toLocaleTimeString('en-US', { hour: '2-digit', minute: '2-digit', hour12: false });
    return this.add({
      type: 'session',
      title: `Session ${date} ${time}`,
      content,
      tags: ['auto-saved', ...tags],
      source: 'stop-hook',
    });
  }

  summary(): string {
    const counts = this.db.prepare(`
      SELECT type, COUNT(*) as count FROM memory WHERE status = 'active' GROUP BY type
    `).all() as { type: string; count: number }[];

    const superseded = (this.db.prepare(`SELECT COUNT(*) as n FROM memory WHERE status = 'superseded'`).get() as any).n;

    if (counts.length === 0) return 'No project memory yet.';

    const lines = counts.map(r => `  ${r.type}: ${r.count} entries`);
    const total = counts.reduce((s, r) => s + r.count, 0);
    const historyNote = superseded > 0 ? `\n  (+ ${superseded} superseded entries in history)` : '';
    return `Project Memory Summary (${total} active):\n${lines.join('\n')}${historyNote}`;
  }

  /**
   * Generate a compact markdown snapshot of all active memory.
   * Used for injecting into CLAUDE.md / system prompts.
   */
  generateContextMd(): string {
    const entries = this.list(200);
    if (entries.length === 0) {
      return '<!-- No project memory yet. Use context-keeper MCP tools to start building memory. -->';
    }

    const grouped = entries.reduce((acc, e) => {
      if (!acc[e.type]) acc[e.type] = [];
      acc[e.type].push(e);
      return acc;
    }, {} as Record<string, MemoryEntry[]>);

    const order: MemoryEntry['type'][] = ['workspace', 'decision', 'pattern', 'convention', 'gotcha', 'context', 'note', 'rejected', 'session'];
    const sections = order
      .filter(t => grouped[t]?.length)
      .map(t => {
        const header = `### ${t.charAt(0).toUpperCase() + t.slice(1)}s`;
        const items = grouped[t].map(e =>
          `- **${e.title}** [id:${e.id}]: ${e.content.split('\n')[0]}${e.tags.length ? ` *(${e.tags.join(', ')})*` : ''}`
        ).join('\n');
        return `${header}\n${items}`;
      });

    return sections.join('\n\n');
  }

  close(): void {
    this.db.close();
    this.globalDb?.close();
    this.workspaceDb?.close();
  }

  private deserialize(row: any): MemoryEntry {
    return {
      ...row,
      tags: JSON.parse(row.tags || '[]'),
      supersedes_id: row.supersedes_id ?? null,
      change_reason: row.change_reason ?? null,
    };
  }
}

export function resolveDbPath(projectRoot: string): string {
  return path.join(projectRoot, '.context', 'context.db');
}

export function resolveGlobalDbPath(): string {
  return path.join(os.homedir(), '.context-keeper', 'global.db');
}

const WORKSPACE_MARKERS = ['lerna.json', 'pnpm-workspace.yaml', 'nx.json', 'turbo.json'];

export function resolveWorkspaceRoot(projectRoot: string): string | null {
  let dir = path.dirname(projectRoot);
  for (let i = 0; i < 3; i++) {
    if (!dir || dir === projectRoot) return null;
    // manual sentinel file — drop a `.context-keeper-workspace` at your monorepo root
    if (fs.existsSync(path.join(dir, '.context-keeper-workspace'))) return dir;
    // standard monorepo markers
    for (const m of WORKSPACE_MARKERS) {
      if (fs.existsSync(path.join(dir, m))) return dir;
    }
    // package.json with "workspaces" field
    const pkgPath = path.join(dir, 'package.json');
    if (fs.existsSync(pkgPath)) {
      try {
        if (JSON.parse(fs.readFileSync(pkgPath, 'utf8')).workspaces) return dir;
      } catch { /* ignore parse errors */ }
    }
    const parent = path.dirname(dir);
    if (parent === dir) return null;
    dir = parent;
  }
  return null;
}

export function resolveWorkspaceDbPath(projectRoot: string): string | null {
  const root = resolveWorkspaceRoot(projectRoot);
  return root ? path.join(root, '.context', 'workspace.db') : null;
}

File: go/internal/storage/storage.go
package storage

import (
	"database/sql"
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"strings"

	_ "modernc.org/sqlite"
)

// ValidTypes lists all allowed memory entry types.
var ValidTypes = []string{
	"decision", "convention", "gotcha", "context",
	"note", "session", "rejected", "pattern", "workspace",
}

// WorkspaceMirrorTypes are mirrored to workspace/global DBs.
var WorkspaceMirrorTypes = map[string]bool{
	"decision": true, "pattern": true, "workspace": true, "rejected": true,
}

// Entry represents a memory record.
type Entry struct {
	ID           int64
	Type         string
	Title        string
	Content      string
	Tags         []string
	Source       *string
	Status       string
	SupersedesID *int64
	ChangeReason *string
	CreatedAt    string
	UpdatedAt    string
}

// SearchResult wraps Entry with a relevance score.
type SearchResult struct {
	Entry
	Score float64
}

// DecisionHistory contains the current version and all superseded ancestors.
type DecisionHistory struct {
	Current Entry
	History []Entry
}

// Config holds database paths and project root.
type Config struct {
	DbPath        string
	ProjectRoot   string
	GlobalDbPath  string
	WorkspaceDbPath string
}

// Storage manages SQLite databases for a project.
type Storage struct {
	cfg         Config
	db          *sql.DB
	globalDb    *sql.DB
	workspaceDb *sql.DB
	onAdd       func(Entry)
}

// OnAdd registers a callback invoked after every successful Add.
func (s *Storage) OnAdd(fn func(Entry)) { s.onAdd = fn }

// New opens (and initialises) all configured databases.
func New(cfg Config) (*Storage, error) {
	s := &Storage{cfg: cfg}

	db, err := openDB(cfg.DbPath)
	if err != nil {
		return nil, fmt.Errorf("open project db: %w", err)
	}
	s.db = db

	if err := migrate(db); err != nil {
		return nil, fmt.Errorf("migrate project db: %w", err)
	}

	if cfg.GlobalDbPath != "" {
		gdb, err := openDB(cfg.GlobalDbPath)
		if err == nil {
			s.globalDb = gdb
			_ = migrate(gdb)
		}
	}

	if cfg.WorkspaceDbPath != "" {
		wdb, err := openDB(cfg.WorkspaceDbPath)
		if err == nil {
			s.workspaceDb = wdb
			_ = migrate(wdb)
		}
	}

	return s, nil
}

// DB returns the underlying project database connection (for markdown generation).
func (s *Storage) DB() *sql.DB {
	return s.db
}

// WorkspaceDB returns the workspace database connection, or nil.
func (s *Storage) WorkspaceDB() *sql.DB {
	return s.workspaceDb
}

// Close closes all open database connections.
func (s *Storage) Close() {
	if s.db != nil {
		s.db.Close()
	}
	if s.globalDb != nil {
		s.globalDb.Close()
	}
	if s.workspaceDb != nil {
		s.workspaceDb.Close()
	}
}

// Add inserts a new entry and mirrors it to global/workspace DBs.
func (s *Storage) Add(e Entry) (Entry, error) {
	id, err := insertEntry(s.db, e)
	if err != nil {
		return Entry{}, err
	}
	result, err := s.GetByID(id)
	if err != nil {
		return Entry{}, err
	}

	// Mirror to global DB (tag with project root as source)
	if s.globalDb != nil {
		mirrored := e
		mirrored.Source = &s.cfg.ProjectRoot
		_, _ = insertEntry(s.globalDb, mirrored)
	}

	// Mirror workspace-relevant types to workspace DB
	if s.workspaceDb != nil && WorkspaceMirrorTypes[e.Type] {
		mirrored := e
		mirrored.Source = &s.cfg.ProjectRoot
		_, _ = insertEntry(s.workspaceDb, mirrored)
	}

	if s.onAdd != nil {
		s.onAdd(result)
	}
	return result, nil
}

// UpdateDecision marks the old entry superseded and creates a new active entry.
func (s *Storage) UpdateDecision(supersedesID int64, title, content string, tags []string, changeReason string) (Entry, error) {
	old, err := s.GetByID(supersedesID)
	if err != nil {
		return Entry{}, fmt.Errorf("entry #%d not found: %w", supersedesID, err)
	}

	tx, err := s.db.Begin()
	if err != nil {
		return Entry{}, err
	}
	defer tx.Rollback()

	_, err = tx.Exec(`UPDATE memory SET status = 'superseded', updated_at = datetime('now') WHERE id = ?`, supersedesID)
	if err != nil {
		return Entry{}, err
	}
	if err := tx.Commit(); err != nil {
		return Entry{}, err
	}

	sid := supersedesID
	newEntry := Entry{
		Type:         old.Type,
		Title:        title,
		Content:      content,
		Tags:         tags,
		SupersedesID: &sid,
		ChangeReason: &changeReason,
	}
	return s.Add(newEntry)
}

// GetByID fetches a single entry by ID.
func (s *Storage) GetByID(id int64) (Entry, error) {
	row := s.db.QueryRow(`SELECT * FROM memory WHERE id = ?`, id)
	return scanEntry(row)
}

// GetByIDs fetches multiple entries by ID.
func (s *Storage) GetByIDs(ids []int64) ([]Entry, error) {
	results := make([]Entry, 0, len(ids))
	for _, id := range ids {
		e, err := s.GetByID(id)
		if err != nil {
			continue
		}
		results = append(results, e)
	}
	return results, nil
}

// List returns up to limit active entries.
func (s *Storage) List(limit int) ([]Entry, error) {
	rows, err := s.db.Query(`SELECT * FROM memory WHERE status = 'active' ORDER BY updated_at DESC LIMIT ?`, limit)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	return scanEntries(rows)
}

// ListSince returns active entries created after the given datetime string.
func (s *Storage) ListSince(since string) ([]Entry, error) {
	rows, err := s.db.Query(
		`SELECT * FROM memory WHERE status = 'active' AND created_at >= ? ORDER BY created_at ASC`,
		since,
	)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	return scanEntries(rows)
}

// ListByType returns all active entries of a given type.
func (s *Storage) ListByType(typ string) ([]Entry, error) {
	rows, err := s.db.Query(
		`SELECT * FROM memory WHERE status = 'active' AND type = ? ORDER BY updated_at DESC`,
		typ,
	)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	return scanEntries(rows)
}

// ListAll returns all active entries, optionally filtered by type.
func (s *Storage) ListAll(typ string) ([]Entry, error) {
	if typ == "" || typ == "all" {
		return s.List(200)
	}
	return s.ListByType(typ)
}

// ListProjects returns distinct project names (basename of source) from the global DB.
func (s *Storage) ListProjects() []string {
	if s.globalDb == nil {
		return nil
	}
	rows, err := s.globalDb.Query(
		`SELECT DISTINCT source FROM memory WHERE status='active' AND source IS NOT NULL AND source != '' ORDER BY source`)
	if err != nil {
		return nil
	}
	defer rows.Close()
	var projects []string
	for rows.Next() {
		var src string
		if err := rows.Scan(&src); err == nil && src != "" {
			projects = append(projects, src)
		}
	}
	return projects
}

// ListGlobalByProject returns all active entries for a given project source path.
// If source is empty, returns all entries across all projects from the global DB.
func (s *Storage) ListGlobalByProject(source, typ string) ([]Entry, error) {
	if s.globalDb == nil {
		return s.ListAll(typ)
	}
	var (
		args  []interface{}
		where = "status='active'"
	)
	if source != "" {
		where += " AND source=?"
		args = append(args, source)
	}
	if typ != "" && typ != "all" {
		where += " AND type=?"
		args = append(args, typ)
	}
	rows, err := s.globalDb.Query(
		`SELECT id,type,title,content,tags,source,status,supersedes_id,change_reason,created_at,updated_at
		 FROM memory WHERE `+where+` ORDER BY updated_at DESC LIMIT 500`, args...)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	return scanEntries(rows)
}

// GetHistory returns the full decision history chain for an entry.
func (s *Storage) GetHistory(id int64) (*DecisionHistory, error) {
	current, err := s.GetByID(id)
	if err != nil {
		return nil, err
	}

	var history []Entry
	cur := current
	for cur.SupersedesID != nil {
		prev, err := s.GetByID(*cur.SupersedesID)
		if err != nil {
			break
		}
		history = append(history, prev)
		cur = prev
	}

	return &DecisionHistory{Current: current, History: history}, nil
}

// Search runs FTS5 search on the local DB.
func (s *Storage) Search(query string, limit int) ([]SearchResult, error) {
	return ftsSearch(s.db, query, limit, "", "")
}

// SearchGlobal searches the global DB excluding entries from the current project.
func (s *Storage) SearchGlobal(query string, limit int) ([]SearchResult, error) {
	if s.globalDb == nil {
		return nil, nil
	}
	return ftsSearch(s.globalDb, query, limit, "exclude_source", s.cfg.ProjectRoot)
}

// SearchWorkspace searches the workspace DB.
func (s *Storage) SearchWorkspace(query string, limit int) ([]SearchResult, error) {
	if s.workspaceDb == nil {
		return nil, nil
	}
	return ftsSearch(s.workspaceDb, query, limit, "", "")
}

// SearchAll merges results from local, workspace, and global DBs.
func (s *Storage) SearchAll(query string, limit int) ([]SearchResult, error) {
	local, _ := s.Search(query, limit)
	workspace, _ := s.SearchWorkspace(query, limit)
	global, _ := s.SearchGlobal(query, limit)

	merged := append(local, workspace...)
	merged = append(merged, global...)

	// Sort by score descending
	for i := 1; i < len(merged); i++ {
		for j := i; j > 0 && merged[j].Score > merged[j-1].Score; j-- {
			merged[j], merged[j-1] = merged[j-1], merged[j]
		}
	}
	if len(merged) > limit {
		merged = merged[:limit]
	}
	return merged, nil
}

// AddSessionSummary saves a session-type entry summarising entries created this session.
func (s *Storage) AddSessionSummary(entries []Entry) error {
	if len(entries) == 0 {
		return nil
	}

	var sb strings.Builder
	fmt.Fprintf(&sb, "%d entries saved this session:\n", len(entries))
	for _, e := range entries {
		firstLine := e.Content
		if idx := strings.Index(firstLine, "\n"); idx != -1 {
			firstLine = firstLine[:idx]
		}
		if len(firstLine) > 80 {
			firstLine = firstLine[:80] + "…"
		}
		fmt.Fprintf(&sb, "- [%s] %s: %s\n", e.Type, e.Title, firstLine)
	}

	source := "stop-hook"
	_, err := s.Add(Entry{
		Type:    "session",
		Title:   fmt.Sprintf("Session %s", entries[0].CreatedAt[:16]),
		Content: sb.String(),
		Tags:    []string{"auto-saved"},
		Source:  &source,
	})
	return err
}

// ── helpers ────────────────────────────────────────────────────────────────────

func openDB(dbPath string) (*sql.DB, error) {
	dir := filepath.Dir(dbPath)
	if err := os.MkdirAll(dir, 0755); err != nil {
		return nil, err
	}
	db, err := sql.Open("sqlite", dbPath)
	if err != nil {
		return nil, err
	}
	db.SetMaxOpenConns(1) // SQLite concurrency: WAL allows concurrent reads but one writer
	if _, err := db.Exec(`PRAGMA journal_mode = WAL`); err != nil {
		return nil, err
	}
	if _, err := db.Exec(schema); err != nil {
		return nil, err
	}
	return db, nil
}

func migrate(db *sql.DB) error {
	rows, err := db.Query(`PRAGMA table_info(memory)`)
	if err != nil {
		return err
	}
	defer rows.Close()

	existing := map[string]bool{}
	for rows.Next() {
		var cid int
		var name, typ, notnull, dflt sql.NullString
		var pk int
		if err := rows.Scan(&cid, &name, &typ, &notnull, &dflt, &pk); err != nil {
			continue
		}
		existing[name.String] = true
	}

	needed := map[string]string{
		"status":        `ALTER TABLE memory ADD COLUMN status TEXT NOT NULL DEFAULT 'active'`,
		"supersedes_id": `ALTER TABLE memory ADD COLUMN supersedes_id INTEGER`,
		"change_reason": `ALTER TABLE memory ADD COLUMN change_reason TEXT`,
	}
	for col, stmt := range needed {
		if !existing[col] {
			if _, err := db.Exec(stmt); err != nil {
				return err
			}
		}
	}
	return nil
}

func insertEntry(db *sql.DB, e Entry) (int64, error) {
	tagsJSON, _ := json.Marshal(e.Tags)
	res, err := db.Exec(
		`INSERT INTO memory (type, title, content, tags, source, supersedes_id, change_reason)
		 VALUES (?, ?, ?, ?, ?, ?, ?)`,
		e.Type, e.Title, e.Content, string(tagsJSON), e.Source, e.SupersedesID, e.ChangeReason,
	)
	if err != nil {
		return 0, err
	}
	return res.LastInsertId()
}

func ftsSearch(db *sql.DB, query string, limit int, mode, modeArg string) ([]SearchResult, error) {
	ftsQuery := sanitizeQuery(query)
	if ftsQuery == "" {
		return nil, nil
	}

	var rows *sql.Rows
	var err error

	if mode == "exclude_source" {
		rows, err = db.Query(
			`SELECT m.*, memory_fts.rank AS score
			 FROM memory_fts
			 JOIN memory m ON memory_fts.rowid = m.id
			 WHERE memory_fts MATCH ?
			   AND m.status = 'active'
			   AND (m.source IS NULL OR m.source != ?)
			 ORDER BY rank LIMIT ?`,
			ftsQuery, modeArg, limit,
		)
	} else {
		rows, err = db.Query(
			`SELECT m.*, memory_fts.rank AS score
			 FROM memory_fts
			 JOIN memory m ON memory_fts.rowid = m.id
			 WHERE memory_fts MATCH ?
			   AND m.status = 'active'
			 ORDER BY rank LIMIT ?`,
			ftsQuery, limit,
		)
	}
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	var results []SearchResult
	for rows.Next() {
		var score float64
		var e Entry
		var tagsStr string
		var source, changeReason sql.NullString
		var supersedesID sql.NullInt64

		err := rows.Scan(
			&e.ID, &e.Type, &e.Title, &e.Content, &tagsStr,
			&source, &e.Status, &supersedesID, &changeReason,
			&e.CreatedAt, &e.UpdatedAt, &score,
		)
		if err != nil {
			continue
		}
		_ = json.Unmarshal([]byte(tagsStr), &e.Tags)
		if source.Valid {
			e.Source = &source.String
		}
		if supersedesID.Valid {
			e.SupersedesID = &supersedesID.Int64
		}
		if changeReason.Valid {
			e.ChangeReason = &changeReason.String
		}
		if score < 0 {
			score = -score
		}
		results = append(results, SearchResult{Entry: e, Score: score})
	}
	return results, nil
}

func scanEntry(row *sql.Row) (Entry, error) {
	var e Entry
	var tagsStr string
	var source, changeReason sql.NullString
	var supersedesID sql.NullInt64

	err := row.Scan(
		&e.ID, &e.Type, &e.Title, &e.Content, &tagsStr,
		&source, &e.Status, &supersedesID, &changeReason,
		&e.CreatedAt, &e.UpdatedAt,
	)
	if err != nil {
		return Entry{}, err
	}
	_ = json.Unmarshal([]byte(tagsStr), &e.Tags)
	if source.Valid {
		e.Source = &source.String
	}
	if supersedesID.Valid {
		e.SupersedesID = &supersedesID.Int64
	}
	if changeReason.Valid {
		e.ChangeReason = &changeReason.String
	}
	return e, nil
}

func scanEntries(rows *sql.Rows) ([]Entry, error) {
	var results []Entry
	for rows.Next() {
		var e Entry
		var tagsStr string
		var source, changeReason sql.NullString
		var supersedesID sql.NullInt64

		err := rows.Scan(
			&e.ID, &e.Type, &e.Title, &e.Content, &tagsStr,
			&source, &e.Status, &supersedesID, &changeReason,
			&e.CreatedAt, &e.UpdatedAt,
		)
		if err != nil {
			continue
		}
		_ = json.Unmarshal([]byte(tagsStr), &e.Tags)
		if source.Valid {
			e.Source = &source.String
		}
		if supersedesID.Valid {
			e.SupersedesID = &supersedesID.Int64
		}
		if changeReason.Valid {
			e.ChangeReason = &changeReason.String
		}
		results = append(results, e)
	}
	return results, rows.Err()
}

// sanitizeQuery converts a natural language query into an FTS5 OR-prefix query.
// "storage backend database" → "storage* OR backend* OR database*"
// This ensures any matching word returns results, not just entries containing all words.
func sanitizeQuery(q string) string {
	q = strings.ReplaceAll(q, "'", " ")
	q = strings.ReplaceAll(q, `"`, " ")
	q = strings.ReplaceAll(q, "*", " ")
	q = strings.TrimSpace(q)
	if q == "" {
		return ""
	}
	words := strings.Fields(q)
	terms := make([]string, 0, len(words))
	for _, w := range words {
		if len(w) >= 2 { // skip single-char words
			terms = append(terms, w+"*")
		}
	}
	if len(terms) == 0 {
		return ""
	}
	return strings.Join(terms, " OR ")
}
